### ISSUE TITLE
Grafana 9.0.0: Error updating options: Unexpected token G in JSON at position 0

### ISSUE THREAD

**Issue by auwaerter** [grafana/grafana#50841]
**What happened**: 
When i'm switching datasources inside a dashboard, the response from grafana is:

E.g. this [dashboard](https://grafana.com/grafana/dashboards/11001)

![grafana](https://user-images.githubusercontent.com/6923945/173750564-c6cb3515-8711-4713-be6a-edc4400e9675.png)

With:

```
Request-URL: https://domain.net/api/datasources/2/resources/api/v1/label/namespace/values?start=1655266619&end=1655270219
Method: GET
Code: 500 
Remote-Address: 192.X.X.X:443
Referral url guidelines: strict-origin-when-cross-origin
```

Also with the [POD Overview](https://grafana.com/grafana/dashboards/6781) dashboard:

```
Request-URL: https://domain.net/api/datasources/3/resources/api/v1/series?match%5B%5D=kube_namespace_status_phase%7Bphase%3D%22Active%22%7D&start=1655268656&end=1655270456
Method: GET
Code: 500 
Remote-Address: 192.X.X.X:443
Referral url guidelines: strict-origin-when-cross-origin
```

**What you expected to happen**: 
The same behavior as in grafana 8.5.6. 

**How to reproduce it (as minimally and precisely as possible)**: 
Having different prometheus datasources in grafana 9.0.0 with browser mode and change them in the dashboard.

**Anything else we need to know?**: 
When i'm choosing a different datasource, the list of namespaces seem not to get updated. As well as all the underlaying objects from a namespace as e.g. pods running. When i'm manually picking a namespace, i can see the metrics successfully in my dashboard. 

I suppose a change/adaption is needed for the variables section in grafana.

Pretty much related: https://github.com/grafana/grafana/issues/50798

**Environment**:
- Grafana version: 9.0.0
- Data source type & version: Prometheus v2.36.1 using Browser Access
- OS Grafana is installed on: On-Premise Kubernetes 1.23.6
- User OS & Browser: Windows 10 20H2 (19042.1706) & Chrome 102.0.5005.115

**Comment by styler** [comment]
> Seeing very similar issue, but not just with browser mode for prometheus. It does not seem to pass any custom headers for variable look ups, like X-Scope-OrgID. Guess its not a problem in Grafana Cloud that use a gateway, but when connecting to multi tenant Cortex / Mimir without authentication gateway this is a big problem.
> 
> Seems like it maybe use different way to fetch variable values than data in graph? From what I could see in code it only set the Authorization header and filtered out all others.



Seeing exactly this behavior.

**Comment by aocenas** [comment]
If root cause is an auth error because of missing headers (I assume it returns something that is not json and we try to parse it as json hence the error message) then this PR should solve this https://github.com/grafana/grafana/pull/50894

**Comment by aocenas** [comment]
Ok seems like this was already fixed but the fix did not make it to 9.0.0 https://github.com/grafana/grafana/pull/50408. As it's already in the branch it should be part of 9.0.1 release.

**Comment by auwaerter** [comment]
Seems like it didn't made it into 9.0.1. I'm having the same issue with that version aswell.

**Comment by bytheway** [comment]
I just upgraded to 9.0.1 and it appears to be fixed for me. I'm getting no more errors on the explore pages, the X-Scope-OrgID header appears to be passed appropriately.

**Comment by auwaerter** [comment]
That's weird, so probably i'm facing a different issue here. 

I've tried again with grafana 9.0.1 to fetch a namespace, based on a prometheus datasource and a query. This is how it looks for me:

![variables](https://user-images.githubusercontent.com/6923945/175226971-f591a084-d189-4eab-aa7b-aa54d6a9cad4.gif)

The "Preview of values" is not showing any update.

While the console is throwing:

![browser](https://user-images.githubusercontent.com/6923945/175227265-16122a95-a6eb-4d4d-853a-a4cdae2d80ae.png)

This is working fine with grafana 8.5.6, but not with 9.0.0/9.0.1.

Here the comparison with 8.5.6:

![8-5-6](https://user-images.githubusercontent.com/6923945/175228243-5a1b7ee6-15f9-407f-b9fd-96fa56a9e98e.gif)

**Comment by auwaerter** [comment]
> @auwaerter are you trying that in our hosted Grafana?

No, this is running on a selfhosted k8s environment

**Comment by tesharp** [comment]
We had the same problem in Grafana Cloud even after 9.0.1. After changing the name of the datasource in settings and saving it somehow worked again, also after I changed it back to what it was previously.

Probably changing name forced it to resave it and save it correctly. Maybe some model changes in upgrade..

**Comment by patjlm** [comment]
We face this as well, even after 9.0.1 upgrade.

The datasource is a prometheus with authentication using a `Authorization` header with a `Bearer` token. There is nothing enabled in the prometheus datasource auth configuration section (except skip tls for my tests). The datasource test works.

Looking at some network traces, I noticed that this header value is prepended with a `Basic` auth data:
```
Authorization: Basic ....,Bearer .....
```
On standard queries, only the Bearer token is set.
The Grafana instance is fronted with an oauth proxy, the `Basic` data corresponds to my username from this oauth.

This prom query with the additional `Basic` data triggers an oauth login on our prometheus backend, and so we don't get the expected query result JSON content.

This is visible on variable queries as well as the datasource explore page.

**Comment by patjlm** [comment]
after further testing, it seems that upgrading our oauth2-proxy in front of grafana did the trick. The `Basic` header is not there anymore in grafana queries to prometheus.
For the record, I tested successfully with [quay.io/oauth2-proxy/oauth2-proxy:v7.3.0](https://quay.io/repository/oauth2-proxy/oauth2-proxy)

**Comment by SuperMatt** [comment]
I found that I was able to fix this in my environment by enabling Cortex Ruler. I was originally only running the basics - querier, fronted, store gateway, ingesters, distributors, but after enabling Ruler the problem goes away. You don't need any rules, it just needs to exist.

**Comment by auwaerter** [comment]
Same issue occurs also with 9.0.3, 8.5.9 is working fine. I do not have any proxy before.
```
Failed to load resource: the server responded with a status of 500 ()
runRequest.ts:151 runRequest.catchError {"data":{"message":"Unexpected error"}}
(anonymous) @ runRequest.ts:151
actions.ts:864 SyntaxError: Unexpected token G in JSON at position 0
(anonymous) @ actions.ts:864
reportUnhandledError.js:10 Uncaught SyntaxError: Unexpected token G in JSON at position 0
/api/datasources/3/resources/api/v1/series?match%5B%5D=http_response_timers_count%7Binstance%3D%22namespace.svc.cluster.local%3A80%22%7D&start=1657689186&end=1657861986:1          Failed to load resource: the server responded with a status of 500 ()
runRequest.ts:151 runRequest.catchError {"data":{"message":"Unexpected error"}}
(anonymous) @ runRequest.ts:151
actions.ts:864 SyntaxError: Unexpected token G in JSON at position 0
(anonymous) @ actions.ts:864
reportUnhandledError.js:10 Uncaught SyntaxError: Unexpected token G in JSON at position 0
/api/datasources/3/resources/api/v1/series?match%5B%5D=http_response_timers_count%7Binstance%3D%22namespace.svc.cluster.local%3A80%22%2C%20database%3D%22_all_%22%7D&start=1657689186&end=1657861986:1          Failed to load resource: the server responded with a status of 500 ()
runRequest.ts:151 runRequest.catchError {"data":{"message":"Unexpected error"}}
(anonymous) @ runRequest.ts:151
actions.ts:864 SyntaxError: Unexpected token G in JSON at position 0
(anonymous) @ actions.ts:864
reportUnhandledError.js:10 Uncaught SyntaxError: Unexpected token G in JSON at position 0
```

**Comment by ciceroverneck** [comment]
/reopen 
Same error occurs in 9.0.3.

**Comment by henricbjork2liu** [comment]
We have this issue since 9.0.* update.
I found that the issue seems to be for every query variable that uses "label_values"

We used to have a query like:
`label_values( windows_device_category{category="Students"}, room)`

This stopped working as soon as we updated to v9.0.0, and still persists when we have v9.0.5
We receive error message:

```
Templating [q_site]

Error updating options: JSON.parse: unexpedted character at line 1
column 1 of the JSON data
```

I got it to work with query_result and regex as a workaround:
query: `query_result(windows_device_category{category="Students"})`
regex: `/room="(?<text>[^"]+)/g`

**Comment by harmjanblok** [comment]
@aocenas can this be reopened? The issue still isn't resolved.

To me it seems that 'Server access mode' is used for a `label_values(metric, label)` query in a Variable definition, however the actual prometheus datasource specifies (the deprecated) 'Browser access mode' (HTTP Method: Post).

I've compared two datasources, both datasources can be reached from the browser. However only one datasource can be reached by the Grafana instance directly.

As @henricbjork2liu writes, the error doesn't occur when using the `query_result(query)` query.

I've tested with Grafana 9.0.7.

**Comment by auwaerter** [comment]
Update

Still the same behavior with grafana 9.1.2

![image](https://user-images.githubusercontent.com/6923945/187597315-9d373849-c525-4bcd-b42e-c3d9dbb8941b.png)

grafana version 8.5.11 instead is working fine.

**Comment by auwaerter** [comment]
@marefr yes exactly, even if the prometheus url is accessible only locally i'm using basic auth (browser access mode) to access the ds.

**Comment by marefr** [comment]
@itsmylife have any input on this? Seems like direct/browser access support is being removed in upcoming Grafana v9.2.0 via #50162?

**Comment by auwaerter** [comment]
It is working with grafana 9.1.5 by switching to server access mode (also: [proxy](https://github.com/grafana/helm-charts/blob/main/charts/grafana/values.yaml#L510)), but with a noticeable slower response in the UI compared to the browser mode.

Also it is tricky to make an external prometheus instance accessible from a grafana container running in another cluster. This is why i was prefering browser mode, even if you have to take care of CORS etc.

**Comment by itsmylife** [comment]
I think we (me and @marefr) have a reproduction and potential fix for this. I am testing the fix and will have more information (hopefully good) soon.

**Comment by itsmylife** [comment]
The fix will be in `v9.1.6` and it will be released today.

**Comment by auwaerter** [comment]
Good Morning @itsmylife &  @marefr, i can also confirm grafana `v9.1.6` is working fine with that fix! Thanks a lot!
